File: _CodeTopics/LeetCode/000086_Partition_List/interview_000086.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def partition(self, head, x):
        """
        :type head: ListNode
        :type x: int
        :rtype: ListNode
        """

        small = curr_small = ListNode(0)
        large = curr_large = ListNode(0)

        while head:
            if head.val < x:
                curr_small.next = head
                curr_small = curr_small.next
                head = head.next
            else:
                curr_large.next = head
                curr_large = curr_large.next
                head = head.next

        # deal with the last node of large list and 
        # concatenate the two to obtain the final result 
        curr_large.next = None
        curr_small.next = large.next
        
        return small.next
"""
https://leetcode-cn.com/submissions/detail/54891785/

166 / 166 个通过测试用例
状态：通过
执行用时: 40 ms
内存消耗: 12.8 MB
"""

File: _CodeTopics/LeetCode/1-200/000086_Partition_List/interview_000086.c
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */


struct ListNode* partition(struct ListNode* head, int x){
    //if (NULL == head){
    //    return -1;
    //}

    struct ListNode* small = (struct ListNode *)malloc(sizeof(struct ListNode));
    struct ListNode* large = (struct ListNode *)malloc(sizeof(struct ListNode));
    struct ListNode* curr_small = small;
    struct ListNode* curr_large = large;

    while (head != NULL){
        if (head->val < x){
            curr_small->next = head;
            curr_small = curr_small->next;
        }
        else{
            curr_large->next = head;
            curr_large = curr_large->next;
        }
        head = head->next;
    }

    curr_large->next = NULL;
    curr_small->next = large->next;
    return small->next;
}
"""
https://leetcode-cn.com/submissions/detail/54908515/

166 / 166 个通过测试用例
状态：通过
执行用时: 4 ms
内存消耗: 5.5 MB
"""
